pkg/ingester/flush.go

```diff
@@ -78,6 +78,13 @@ func (i *Ingester) sweepInstance(instance *instance, immediate bool) {
 		i.sweepStream(instance, stream, immediate)
 		i.removeFlushedChunks(instance, stream)
 	}
+
+	if len(instance.streams) == 0 {
+		i.instancesMtx.Lock()
+		delete(i.instances, instance.instanceID)
+		i.instancesMtx.Unlock()
+		memoryTenants.Dec()
+	}
```

Comment:
I'd say this should go in `sweepUsers`?

Reply:
Actually, looking at it I'm not sure we can take the instancesMtx whilst holding the streamsMtx, and this might deadlock with the write path.  Thinking...

Reply:
Okay it won't dead lock (we release instancesMtx before taking streamsMtx on the write path), but it could lead to use dropping a reference to the instance, if the GC collection happens after the write path has got the instance but before its take streamsMtx.  And you've got a race as you're not checking the map size under the lock.